src/main/java/com/yelp/nrtsearch/server/luceneserver/QueryNodeMapper.java

```diff
@@ -643,4 +636,36 @@ private Query getConstantScoreQuery(
     Query filterQuery = getQuery(constantScoreQueryGrpc.getFilter(), indexState, docLookup);
     return new ConstantScoreQuery(filterQuery);
   }
+
+  private SpanNearQuery getSpanNearQuery(
+      com.yelp.nrtsearch.server.grpc.SpanNearQuery protoSpanNearQuery) {
+    List<SpanQuery> clauses = new ArrayList<>();
+
+    for (com.yelp.nrtsearch.server.grpc.SpanQuery protoClause :
+        protoSpanNearQuery.getClausesList()) {
+      SpanQuery luceneClause = convertToLuceneSpanQuery(protoClause);
+      if (luceneClause != null) {
+        clauses.add(luceneClause);
+      }
+    }
+
+    SpanQuery[] clausesArray = clauses.toArray(new SpanQuery[0]);
+    int slop = protoSpanNearQuery.getSlop();
+    boolean inOrder = protoSpanNearQuery.getInOrder();
+
+    return new SpanNearQuery(clausesArray, slop, inOrder);
+  }
+
+  private SpanQuery convertToLuceneSpanQuery(
+      com.yelp.nrtsearch.server.grpc.SpanQuery protoSpanQuery) {
+    com.yelp.nrtsearch.server.grpc.TermQuery termQuery = protoSpanQuery.getSpanTermQuery();
+    if (termQuery != null) {
+      return new SpanTermQuery(new Term(termQuery.getField(), termQuery.getTextValue()));
+    }
+    com.yelp.nrtsearch.server.grpc.SpanNearQuery spanNearQuery = protoSpanQuery.getSpanNearQuery();
+    if (spanNearQuery != null) {
+      return getSpanNearQuery(spanNearQuery);
+    }
+    throw new IllegalArgumentException("Unsupported Span Query: " + protoSpanQuery);
```

Comment:
This can be done with a switch statement using the field case value. Something similar to https://github.com/Yelp/nrtsearch/blob/master/src/main/java/com/yelp/nrtsearch/server/luceneserver/QueryNodeMapper.java#L137

Reply:
Good suggestions. I applied the necessary changes.